File: src/services/customPracticesService.js
import { v4 as uuidv4 } from 'uuid';

const STORAGE_KEY_PRACTICES = 'custom_practices';
const STORAGE_KEY_FOLDERS = 'practice_folders';

export class CustomPracticesService {
  // Get all practices
  static getPractices() {
    try {
      const practices = localStorage.getItem(STORAGE_KEY_PRACTICES);
      return practices ? JSON.parse(practices) : [];
    } catch (error) {
      console.error('Error loading practices:', error);
      return [];
    }
  }

  // Get all folders
  static getFolders() {
    try {
      const folders = localStorage.getItem(STORAGE_KEY_FOLDERS);
      return folders ? JSON.parse(folders) : [];
    } catch (error) {
      console.error('Error loading folders:', error);
      return [];
    }
  }

  // Create a new practice
  static async createPractice({ name, description, duration, color, folderId = null }) {
    const practices = this.getPractices();
    const newPractice = {
      id: uuidv4(),
      name,
      description,
      duration,
      color,
      folderId,
      createdAt: new Date().toISOString(),
    };
    practices.push(newPractice);
    localStorage.setItem(STORAGE_KEY_PRACTICES, JSON.stringify(practices));
    return newPractice;
  }

  // Create a new folder
  static async createFolder({ name, color = '#E8D5F2', icon = '📁' }) {
    const folders = this.getFolders();
    const newFolder = {
      id: uuidv4(),
      name,
      color,
      icon,
      createdAt: new Date().toISOString(),
    };
    folders.push(newFolder);
    localStorage.setItem(STORAGE_KEY_FOLDERS, JSON.stringify(folders));
    return newFolder;
  }

  // Delete a practice
  static async deletePractice(practiceId) {
    const practices = this.getPractices();
    const filtered = practices.filter(p => p.id !== practiceId);
    localStorage.setItem(STORAGE_KEY_PRACTICES, JSON.stringify(filtered));
  }

  // Delete a folder (and clear folderId from practices in it)
  static async deleteFolder(folderId) {
    // Remove folder
    const folders = this.getFolders();
    const filtered = folders.filter(f => f.id !== folderId);
    localStorage.setItem(STORAGE_KEY_FOLDERS, JSON.stringify(filtered));

    // Clear folderId from practices
    const practices = this.getPractices();
    const updated = practices.map(p => 
      p.folderId === folderId ? { ...p, folderId: null } : p
    );
    localStorage.setItem(STORAGE_KEY_PRACTICES, JSON.stringify(updated));
  }

  // Move practice to a folder
  static async movePracticeToFolder(practiceId, folderId) {
    const practices = this.getPractices();
    const updated = practices.map(p =>
      p.id === practiceId ? { ...p, folderId } : p
    );
    localStorage.setItem(STORAGE_KEY_PRACTICES, JSON.stringify(updated));
  }

  // Get practices by folder
  static getPracticesByFolder(folderId) {
    const practices = this.getPractices();
    return practices.filter(p => p.folderId === folderId);
  }

  // Get uncategorized practices
  static getUncategorizedPractices() {
    const practices = this.getPractices();
    return practices.filter(p => !p.folderId);
  }
}

File: mobile-app-flutter/lib/services/custom_practices_service.dart
import 'dart:convert';
import 'package:shared_preferences/shared_preferences.dart';
import '../models/custom_practice.dart';
import 'package:uuid/uuid.dart';

class CustomPracticesService {
  static const String _practicesKey = 'custom_practices';
  static const String _foldersKey = 'practice_folders';
  final _uuid = const Uuid();

  // Get all practices
  Future<List<CustomPractice>> getPractices() async {
    final prefs = await SharedPreferences.getInstance();
    final practicesJson = prefs.getString(_practicesKey);
    
    if (practicesJson == null) return [];
    
    final List<dynamic> decoded = jsonDecode(practicesJson);
    return decoded.map((json) => CustomPractice.fromJson(json)).toList();
  }

  // Get all folders
  Future<List<PracticeFolder>> getFolders() async {
    final prefs = await SharedPreferences.getInstance();
    final foldersJson = prefs.getString(_foldersKey);
    
    if (foldersJson == null) return [];
    
    final List<dynamic> decoded = jsonDecode(foldersJson);
    return decoded.map((json) => PracticeFolder.fromJson(json)).toList();
  }

  // Create new practice
  Future<CustomPractice> createPractice({
    required String name,
    required String description,
    required String duration,
    required String color,
    String? folderId,
  }) async {
    final practice = CustomPractice(
      id: _uuid.v4(),
      name: name,
      description: description,
      duration: duration,
      color: color,
      folderId: folderId,
      createdAt: DateTime.now(),
    );

    final practices = await getPractices();
    practices.add(practice);
    await _savePractices(practices);

    return practice;
  }

  // Create new folder
  Future<PracticeFolder> createFolder({
    required String name,
    required String color,
    required String icon,
  }) async {
    final folder = PracticeFolder(
      id: _uuid.v4(),
      name: name,
      color: color,
      icon: icon,
      createdAt: DateTime.now(),
    );

    final folders = await getFolders();
    folders.add(folder);
    await _saveFolders(folders);

    return folder;
  }

  // Delete practice
  Future<void> deletePractice(String id) async {
    final practices = await getPractices();
    practices.removeWhere((p) => p.id == id);
    await _savePractices(practices);
  }

  // Delete folder (and move practices to uncategorized)
  Future<void> deleteFolder(String id) async {
    final folders = await getFolders();
    folders.removeWhere((f) => f.id == id);
    await _saveFolders(folders);

    // Move practices to uncategorized
    final practices = await getPractices();
    final updatedPractices = practices.map((p) {
      if (p.folderId == id) {
        return CustomPractice(
          id: p.id,
          name: p.name,
          description: p.description,
          duration: p.duration,
          color: p.color,
          folderId: null,
          createdAt: p.createdAt,
        );
      }
      return p;
    }).toList();
    await _savePractices(updatedPractices);
  }

  // Update practice folder
  Future<void> movePracticeToFolder(String practiceId, String? folderId) async {
    final practices = await getPractices();
    final updatedPractices = practices.map((p) {
      if (p.id == practiceId) {
        return CustomPractice(
          id: p.id,
          name: p.name,
          description: p.description,
          duration: p.duration,
          color: p.color,
          folderId: folderId,
          createdAt: p.createdAt,
        );
      }
      return p;
    }).toList();
    await _savePractices(updatedPractices);
  }

  // Get practices by folder
  Future<List<CustomPractice>> getPracticesByFolder(String? folderId) async {
    final practices = await getPractices();
    return practices.where((p) => p.folderId == folderId).toList();
  }

  // Private: Save practices
  Future<void> _savePractices(List<CustomPractice> practices) async {
    final prefs = await SharedPreferences.getInstance();
    final json = jsonEncode(practices.map((p) => p.toJson()).toList());
    await prefs.setString(_practicesKey, json);
  }

  // Private: Save folders
  Future<void> _saveFolders(List<PracticeFolder> folders) async {
    final prefs = await SharedPreferences.getInstance();
    final json = jsonEncode(folders.map((f) => f.toJson()).toList());
    await prefs.setString(_foldersKey, json);
  }
}
